plutus-core/testlib/PlutusIR/Generators/QuickCheck/ShrinkTerms.hs

```diff
@@ -275,22 +269,34 @@ shrinkTypedTerm tyctx0 ctx0 (ty0, tm0) = concat
           | TyFun _ _ b <- [ty]
           ]
 
-        Apply _ fun arg | Right argTy <- inferTypeInContext tyctx ctx arg ->
-          -- Drop substerms
-          [(argTy, arg), (TyFun () argTy ty, fun)]
+        Apply _ fun arg -> case inferTypeInContext tyctx ctx arg of
+          Right argTy ->
+            -- Drop substerms
+            [ -- Prefer to keep the same type, but...
+              (ty, fixupTerm tyctx ctx tyctx ctx ty arg)
+            , (ty, fixupTerm tyctx ctx tyctx ctx ty fun)
```

Comment:
These two do allow us to get better shrinks. Makes me wonder how many other such opportunities we have unrealized...

Reply:
I don't understand this - what do they do?

Reply:
Made it

```
            [ -- Prefer to keep the same type, so that we don't need to change types elsewhere,
              -- which may remove or cover up the culprit.
              (ty, fixupTerm tyctx ctx tyctx ctx ty arg)
            , (ty, fixupTerm tyctx ctx tyctx ctx ty fun)
            , -- But support type-changing shrinking as well in case it's 'fixupTerm' that removes
              -- or covers up the culprit.
              (argTy, arg)
            , (TyFun () argTy ty, fun)
            ]
```

Hopefully that's clearer.

Reply:
I still don't understand what this is trying to do exactly. `ty` seems to be the type of `Apply fun arg`. Are we somehow "fixing up" `arg` so that it has type `ty`? How does that work?

Reply:
> Are we somehow "fixing up" arg so that it has type ty?

That's the intention, yes.

> How does that work?

Badly it seems, I didn't get to improving it or even reviewing it properly, I think currently it just generates a minimal new term. There's also `fixupTerm_` that seems to be more clever, but I didn't review it either. Term generators and shrinkers were half-reviewed and almost zero-improved.

Reply:
Note that a function can be applied to an argument of the same type that the function returns. `f :: Integer -> Integer` for example. So if `arg` type checks against `ty`, then I think we'll get an ideal shrink. But again, I didn't even get to review the whole thing properly.

Reply:
Interesting. Are you sure the fixed up term isn't actually bigger than `Apply fun arg`? Otherwise we'd be shrinking in the wrong direction.

Reply:
> Interesting. Are you sure the fixed up term isn't actually bigger than `Apply fun arg`? Otherwise we'd be shrinking in the wrong direction.

No, but at least we have a test that is supposed to detect shrinking loops. We discussed it a lot back when PIR generators were getting reviewed and I did ask Max for a definitive Note multiple times (e.g. [here](https://github.com/IntersectMBO/plutus/pull/4559#issuecomment-1240746814)), but we decided to do it as a follow-up after [this](https://github.com/IntersectMBO/plutus/pull/4559#issuecomment-1241882441) response and as usual the follow-up never happened.